app/src/test/kotlin/ru/yandex/loginapp/LoginViewModelTest.kt

```diff
@@ -0,0 +1,72 @@
+package ru.yandex.loginapp
+
+import junit.framework.TestCase.assertEquals
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.StandardTestDispatcher
+import kotlinx.coroutines.test.TestDispatcher
+import kotlinx.coroutines.test.advanceUntilIdle
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.runCurrent
+import kotlinx.coroutines.test.runTest
+import kotlinx.coroutines.test.setMain
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import ru.yandex.loginapp.LoginScreenState.EmptyFieldsError
+
+@OptIn(ExperimentalCoroutinesApi::class)
+class LoginViewModelTest {
+
+    lateinit var viewModel: LoginViewModel
+    val dispatcher: TestDispatcher = StandardTestDispatcher()
+
+    @Before
+    fun setUp() {
+        viewModel = LoginViewModel()
+        Dispatchers.setMain(dispatcher)
+    }
+
+    @After
+    fun tearDown() {
+        Dispatchers.resetMain()
+    }
+
+    @Test
+    fun `login with empty fields sets EmptyFieldsError` () = runTest {
+        val expected = EmptyFieldsError
+        viewModel.login("", "")
+        val actual = viewModel.state.value
+        assertEquals(actual, expected)
+    }
+
+    @Test
+    fun `login with invalid email sets EmailValidationError`() = runTest {
+        val expected = LoginScreenState.EmailValidationError
+        viewModel.login("ya.sboevandreyyandex.ru", "12345")
+        val actual = viewModel.state.value
+        assertEquals(actual, expected)
+    }
+
+    @Test
+    fun `login with valid data sets Loading`() {
+        runTest(dispatcher) {
```

Comment:
Здесь аналогично другим тестам лучше написать так:
```suggestion
    fun `login with valid data sets Loading`() = runTest(dispatcher) {
```